dns_sd.c

```diff
@@ -0,0 +1,172 @@
+	int i;
+
+	d = *ddata;
+
+	if (n == 0) {
+		tdata = d->next;
+		dnssd_free_discovery_data(d);
+		d = tdata;
+	} else {
+		for (i = 0, ndata = d; ndata->next != NULL; ndata = ndata->next) {
+			if (i == n) {
+				/* Could be NULL or node, both are OK */
+				tdata = ndata->next;
+				/* free the node to be removed */
+				dnssd_free_discovery_data(ndata);
+				/* snip it out */
+				ldata->next = tdata;
+				break;
+			}
+			ldata = ndata;
+			i++;
+		}
+	}
+
+	*ddata = d;
+}
+
+/*
+ * remove the ones in the list that you can't connect to
+ * This is sort of silly, but we have seen non-iio devices advertised
+ * and discovered on the network. Oh well....
+ */
+void port_knock_discovery_data(struct dns_sd_discovery_data **ddata)
+{
+	struct dns_sd_discovery_data *d, *ndata;
+	int i, ret;
+
+	d = *ddata;
+	iio_mutex_lock(d->lock);
+	for (i = 0, ndata = d; ndata->next != NULL; ndata = ndata->next) {
+		char port_str[6];
+		struct addrinfo hints, *res, *rp;
+		int fd;
+
+		memset(&hints, 0, sizeof(hints));
+		hints.ai_family = AF_UNSPEC;
+		hints.ai_socktype = SOCK_STREAM;
+
+		iio_snprintf(port_str, sizeof(port_str), "%hu", ndata->port);
+		ret = getaddrinfo(ndata->addr_str, port_str, &hints, &res);
+
+		/* getaddrinfo() returns a list of address structures */
+		if (ret) {
+			DEBUG("Unable to find host ('%s'): %s\n",
+					ndata->hostname,
+					gai_strerror(ret));
+			dnssd_remove_node(&d, i);
+		} else {
+			for (rp = res; rp != NULL; rp = rp->ai_next) {
+				fd = create_socket(rp, DEFAULT_TIMEOUT_MS);
+				if (fd < 0) {
+					DEBUG("Unable to open %s%s socket ('%s:%d' %s)\n",
+							res->ai_family == AF_INET ? "ipv4" : "",
+							res->ai_family == AF_INET6? "ipv6" : "",
+					ndata->hostname, ndata->port, ndata->addr_str);
+					dnssd_remove_node(&d, i);
+				} else {
+					close(fd);
+					DEBUG("Something %s%s at '%s:%d' %s)\n",
+							res->ai_family == AF_INET ? "ipv4" : "",
+							res->ai_family == AF_INET6? "ipv6" : "",
+							ndata->hostname, ndata->port, ndata->addr_str);
+					i++;
+				}
+			}
+		}
+		freeaddrinfo(res);
+	}
+	iio_mutex_unlock(d->lock);
+	*ddata = d;
+
+	return;
+}
+
+void remove_dup_discovery_data(struct dns_sd_discovery_data **ddata)
+{
+	struct dns_sd_discovery_data *d, *ndata, *mdata;
+	int i, j;
+
+	d = *ddata;
+
+	if (!d)
+		return;
+
+	if (!d->next)
+		return;
+
+	iio_mutex_lock(d->lock);
+	for (i = 0, ndata = d; ndata->next != NULL; ndata = ndata->next) {
+		for (j = i + 1, mdata = ndata->next; mdata->next != NULL; mdata = mdata->next) {
+			if (!strcmp(mdata->hostname, ndata->hostname) &&
+					!strcmp(mdata->addr_str, ndata->addr_str)){
+				DEBUG("Removing duplicate in list: '%s'\n",
+						ndata->hostname);
+				dnssd_remove_node(&d, j);
+			}
+			j++;
+		}
+		i++;
+	}
+	iio_mutex_unlock(d->lock);
+
+	*ddata = d;
+}
+
+int dnssd_discover_host(char *addr_str, size_t addr_len, uint16_t *port)
+{
+	struct dns_sd_discovery_data *ddata;
+	int ret = 0;
+
+	ret = dnssd_find_hosts(&ddata);
+
+	if (ddata) {
```

Comment:
we should either check that `ret < 0` or initialize `ddata` to NULL;
on x86 I've seen that values on stack get initialized to zero [most times], but on ARM it can be anything, and then you get weird behavior/crashes;

Reply:
`dnssd_find_hosts()` allocates memory or returns NULL on most failures. (just looking with a better eye now), so yes - you are right.

fixed.